src/shogun2-core/src/main/java/de/terrestris/shogun2/service/AbstractSecuredPersistentObjectService.java

```diff
@@ -0,0 +1,243 @@
+import de.terrestris.shogun2.dao.PermissionCollectionDao;
+import de.terrestris.shogun2.model.SecuredPersistentObject;
+import de.terrestris.shogun2.model.User;
+import de.terrestris.shogun2.model.UserGroup;
+import de.terrestris.shogun2.model.security.Permission;
+import de.terrestris.shogun2.model.security.PermissionCollection;
+
+/**
+ *
+ * @author Nils Bühner
+ * @see AbstractCrudService
+ *
+ */
+public abstract class AbstractSecuredPersistentObjectService<E extends SecuredPersistentObject, D extends GenericHibernateDao<E, Integer>>
+		extends AbstractCrudService<E, D> {
+
+	@Autowired
+	@Qualifier("permissionCollectionService")
+	protected PermissionCollectionService<PermissionCollection, PermissionCollectionDao<PermissionCollection>> permissionCollectionService;
+
+	/**
+	 * Default constructor, which calls the type-constructor
+	 */
+	@SuppressWarnings("unchecked")
+	public AbstractSecuredPersistentObjectService() {
+		this((Class<E>) SecuredPersistentObject.class);
+	}
+
+	/**
+	 * Constructor that sets the concrete entity class for the service.
+	 * Subclasses MUST call this constructor.
+	 */
+	protected AbstractSecuredPersistentObjectService(Class<E> entityClass) {
+		super(entityClass);
+	}
+
+	/**
+	 * This method adds (user) permissions to the passed entity and persists (!)
+	 * the permission collection!
+	 * If no permissions have been set before, they will be created. Otherwise
+	 * the passed permissions will be added to the existing permission
+	 * collection.
+	 *
+	 * @param entity The secured entity
+	 * @param user The user that gets permissions for the entity
+	 * @param permissions The permissions the user gets for the entity
+	 */
+	public void addUserPermissions(E entity, User user, Permission... permissions) {
+		if(entity != null) {
+			// create a set from the passed array
+			final HashSet<Permission> permissionsSet = new HashSet<Permission>(Arrays.asList(permissions));
+
+			if(permissionsSet != null && !permissionsSet.isEmpty()) {
+				// get the existing permission
+				PermissionCollection userPermissionCollection = entity.getUserPermissions().get(user);
+
+				// whether or not we have to persist the entity (which is only the case
+				// if a new permission collection will be created in the next step)
+				boolean persistEntity = false;
+
+				if(userPermissionCollection == null) {
+					// create a new user permission collection and attach it to the user
+					entity.getUserPermissions().put(user, new PermissionCollection(permissionsSet));
+					LOG.debug("Attached a new permission collection for a user: " + permissionsSet);
+
+					// also persist the entity as a new permission collection has been attached
+					persistEntity = true;
+				} else {
+					// add the passed permissions to the the existing permission collection
+					userPermissionCollection.getPermissions().addAll(permissionsSet);
+					LOG.debug("Added the following permissions to an existing permission collection: "
```

Comment:
You aren't really sure if you actually added something, right? Couldn't we compare the `size()` of the `Set`, to see if anything changed at all? And only in such cases actually save?